crates/core/tedge_agent/src/tedge_to_te_converter/converter.rs

```diff
@@ -0,0 +1,153 @@
+        Ok(self.wrap_errors(messages_or_err))
+    }
+}
+
+impl TedgetoTeConverter {
+    pub fn new() -> Self {
+        TedgetoTeConverter {}
+    }
+
+    fn try_convert(
+        &mut self,
+        mut message: MqttMessage,
+    ) -> Result<Vec<tedge_mqtt_ext::Message>, TedgetoTeConverterError> {
+        match message.topic.clone() {
+            topic if topic.name.starts_with("tedge/measurements") => {
+                Ok(self.convert_measurement(&mut message))
+            }
+            topic if topic.name.starts_with("tedge/events") => Ok(self.convert_event(&mut message)),
+            topic if topic.name.starts_with("tedge/alarms") => self.convert_alarm(&mut message),
+            topic if topic.name.starts_with("tedge/health") => {
+                Ok(self.convert_health_status_message(&mut message))
+            }
+            _ => Ok(vec![]),
+        }
+    }
+
+    // tedge/measurements -> te/device/main///m/
+    // tedge/measurements/child -> te/device/child///m/
+    fn convert_measurement(&mut self, message: &mut MqttMessage) -> Vec<tedge_mqtt_ext::Message> {
+        let te_topic = match message.topic.name.split('/').collect::<Vec<_>>()[..] {
+            ["tedge", "measurements"] => Topic::new_unchecked("te/device/main///m/"),
+            ["tedge", "measurements", cid] => {
+                Topic::new_unchecked(format!("te/device/{cid}///m/").as_str())
+            }
+            _ => return vec![],
+        };
+
+        message.topic = te_topic;
+        vec![(message.to_owned())]
+    }
+
+    // tedge/alarms/severity/alarm_type -> te/device/main///a/alarm_type, put severity in payload
+    // tedge/alarms/severity/child/alarm_type ->  te/device/child///a/alarm_type, put severity in payload
+    fn convert_alarm(
+        &mut self,
+        message: &mut MqttMessage,
+    ) -> Result<Vec<MqttMessage>, TedgetoTeConverterError> {
+        let (te_topic, severity) = match message.topic.name.split('/').collect::<Vec<_>>()[..] {
+            ["tedge", "alarms", severity, alarm_type] => (
+                Topic::new_unchecked(format!("te/device/main///a/{alarm_type}").as_str()),
+                severity,
+            ),
+            ["tedge", "alarms", severity, cid, alarm_type] => (
+                Topic::new_unchecked(format!("te/device/{cid}///a/{alarm_type}").as_str()),
+                severity,
+            ),
+            _ => return Ok(vec![]),
+        };
+
+        let mut alarm: ThinEdgeAlarmData = serde_json::from_str(message.payload_str()?)?;
+        alarm.severity = severity.to_owned();
+        Ok(vec![MqttMessage::new(
+            &te_topic,
+            serde_json::to_string(&alarm)?,
+        )
+        .with_qos(message.qos)])
+    }
+
+    // tedge/events/event_type -> te/device/main///e/event_type
+    // tedge/events/child/event_type -> te/device/child///e/event_type
+    fn convert_event(&mut self, message: &mut MqttMessage) -> Vec<tedge_mqtt_ext::Message> {
```

Comment:
I would update the message in place, avoiding to build a clone with `to_owned()`.

```suggestion
    fn convert_event(&mut self, mut message: MqttMessage) -> Vec<tedge_mqtt_ext::Message> {
```

Reply:
fixed